packages/shopify-app-session-storage-drizzle/src/adapters/drizzle-pg.adapter.ts

```diff
@@ -0,0 +1,139 @@
+import {Session} from '@shopify/shopify-api';
+import {SessionStorage} from '@shopify/shopify-app-session-storage';
+import {
+  InferInsertModel,
+  InferSelectModel,
+  desc,
+  eq,
+  inArray,
+} from 'drizzle-orm';
+import {
+  PgDatabase,
+  QueryResultHKT as PgDatabaseQueryResultHKT,
+} from 'drizzle-orm/pg-core';
+
+import {PgSessionTable} from '../schemas/pg.schema';
+
+export class DrizzlePgAdapter<
+  T extends PgDatabase<PgDatabaseQueryResultHKT, TSchema>,
+  TSchema extends {[key: string]: unknown} = {[key: string]: unknown},
+> implements SessionStorage
+{
+  constructor(
+    private readonly db: T,
+    private readonly sessionTable: PgSessionTable,
+  ) {}
+
+  public async storeSession(session: Session): Promise<boolean> {
+    const row = this.sessionToRow(session);
+
+    await this.db
+      .insert(this.sessionTable)
+      .values({...row})
+      .onConflictDoUpdate({
+        target: this.sessionTable.id,
+        set: {...row},
+      });
+
+    return true;
+  }
+
+  public async loadSession(id: string): Promise<Session | undefined> {
+    try {
+      const row = await this.db
+        .select()
+        .from(this.sessionTable)
+        .where(eq(this.sessionTable.id, id));
+
+      if (!row[0]) {
+        return undefined;
+      }
+
+      return this.rowToSession(row[0]);
+    } catch (error) {
+      console.error(error);
+
+      return undefined;
+    }
+  }
+
+  public async deleteSession(id: string): Promise<boolean> {
+    try {
+      await this.db
+        .delete(this.sessionTable)
+        .where(eq(this.sessionTable.id, id));
+    } catch (error) {
+      console.error(error);
+
+      return false;
+    }
+
+    return true;
+  }
+
+  public async deleteSessions(ids: string[]): Promise<boolean> {
+    try {
+      await this.db
+        .delete(this.sessionTable)
+        .where(inArray(this.sessionTable.id, ids));
+
+      return true;
+    } catch (error) {
+      console.error(error);
+
+      return false;
+    }
+  }
+
+  public async findSessionsByShop(shop: string): Promise<Session[]> {
+    const sessions = await this.db
+      .select()
+      .from(this.sessionTable)
+      .where(eq(this.sessionTable.shop, shop))
+      .orderBy(desc(this.sessionTable.expires));
+
+    return sessions.map((session) => this.rowToSession(session));
+  }
+
+  private sessionToRow(session: Session): InferInsertModel<PgSessionTable> {
+    return {
+      id: session.id,
+      shop: session.shop,
+      state: session.state,
+      isOnline: session.isOnline,
+      scope: session.scope,
+      expires: session.expires
+        ? new Date(session.expires).toDateString()
+        : null,
+      accessToken: session.accessToken,
+      userId: session.onlineAccessInfo?.associated_user.id as unknown as bigint,
+    };
+  }
+
+  private rowToSession(row: InferSelectModel<PgSessionTable>): Session {
+    const sessionParams: {[key: string]: boolean | string | number} = {
```

Comment:
This works, but it should also be possible to just create a `Session` with the required values, and then assign the rest to that object.

Reply:
Could you please tell a bit more about this, because I didn't get it, probably some example? Firstly, I declare `sessionParams` with required values and then add optional values.

Reply:
I was just thinking that instead of creating an object here, we could do something like:

```ts
const session = new Session({ <mandatory fields> });

if (row.<field>) {
  session.<field> = row.<field>;
}
```

That would enable us to ensure that the types for the session are being respected as we set the values.

Reply:
Yeah I figured this might happen - I think you can leave this as is and we'll find a broader solution to all of the packages later on.